src/TfaManager.php

```diff
@@ -0,0 +1,310 @@
+      }
+    }
+
+    // Add fallback plugins.
+    foreach ($fallback_plugins as $key) {
+      if (array_key_exists($key, $api)) {
+        $context['plugins']['fallback'][] = $api[$key]['class'];
+      }
+    }
+    // Allow other modules to modify TFA context.
+    \Drupal::moduleHandler()->alter('tfa_context', $context);
+    $this->setContext($account, $context);
+    return $context;
+  }
+
+  /**
+   * Set context for account's TFA process.
+   *
+   * @param $account User account
+   * @param array $context Context array
+   * @see tfa_start_context() for context format
+   */
+  public function setContext($account, $context) {
+    $_SESSION['tfa'][$account->id()] = $context;
+    $_SESSION['tfa'][$account->id()]['uid'] = $account->id();
+    // Clear existing static TFA process.
+    drupal_static_reset('tfa_get_process');
+  }
+
+
+
+  /**
+   * Login submit handler to determine if TFA process is applicable.
+   *
+   * @param array $form
+   * @param array $form_state
+   */
+  public function loginSubmit($form, &$form_state) {
+    // Similar to tfa_user_login() but not required to force user logout.
+    $tfaManager = \Drupal::service('tfa.manager');
+
+    if ($uid = $form_state->get('uid')) {
+      $account = \Drupal::entityManager()->getStorage('user')->load($uid);
+    }
+    else {
+      $account = user_load_by_name($form_state->get('name'));
+    }
+
+    if ($tfa = $tfaManager->getProcess($account)) {
+      if ($account->hasPermission('require tfa') && !$tfaManager->loginComplete($account) && !$tfa->ready()) {
+        drupal_set_message(t('Login disallowed. You are required to setup two-factor authentication. Please contact a site administrator.'), 'error');
+        $form_state['redirect'] = 'user';
+      }
+      elseif (!$tfaManager->loginComplete($account) && $tfa->ready() && !$tfa->loginAllowed($account)) {
+
+        // Restart flood levels, session context, and TFA process.
+        //flood_clear_event('tfa_validate');
+        //flood_register_event('tfa_begin');
+//      $context = tfa_start_context($account);
+//      $tfa = _tfa_get_process($account);
+
+        // $query = drupal_get_query_parameters();
+        if (!empty($form_state->redirect)) {
+          // If there's an existing redirect set it in TFA context and
+          // tfa_form_submit() will extract and set once process is complete.
+          $context['redirect'] = $form_state['redirect'];
+        }
+        unset($_GET['destination']);
+
+        // Begin TFA and set process context.
+        $tfa->begin();
+        $context = $tfa->getContext();
+        $tfaManager->setContext($account, $context);
+
+        $login_hash = $tfaManager->getLoginHash($account);
+        $form_state->setRedirect(
+          'tfa.entry',
+          ['user' => $account->id(),
+            'hash' => $login_hash]
+        //'tfa/' . $account->id() . '/' . $login_hash
+        //array('query' => $query),
+        );
+      }
+      else {
```

Comment:
@nerdstein, @EclipseGc  Any ideas on how to improve this else statement? 

Basically, for this condition we want to login the user normally and not run any TFA code.

We won't know if a user has TFA enabled until they submit the login form, but we can't call the login form ::submitForm method from this point. In D7 they called user_login_submit directly so for now I duplicated the submitForm() code and made it work for D8.

I was thinking if we can leave both submit methods in the form, and instead of using the form_state['redirect'] to send them to the TFA form, we force it to go so the 2nd submit never gets fired. Thoughts?